File: frontend/src/lib/api/bookmarks.ts
import { api } from "./client";

export interface Bookmark {
  id: string;
  live_set_id: string;
  live_set_slug: string | null;
  live_set_title: string | null;
  position_seconds: number | null;
  label: string | null;
  created_at: string;
}

interface BookmarksList {
  items: Bookmark[];
}

export const listForSet = async (slug: string): Promise<Bookmark[]> => {
  const r = await api<BookmarksList>(`/api/sets/${encodeURIComponent(slug)}/bookmarks`);
  return r.items;
};

export const createBookmark = (
  slug: string,
  position_seconds: number | null,
  label: string | null,
) =>
  api<Bookmark>(`/api/sets/${encodeURIComponent(slug)}/bookmarks`, {
    method: "POST",
    body: JSON.stringify({ position_seconds, label }),
  });

export const deleteBookmark = (slug: string, id: string) =>
  api<void>(`/api/sets/${encodeURIComponent(slug)}/bookmarks/${id}`, {
    method: "DELETE",
  });

export const listMyBookmarks = async (): Promise<Bookmark[]> => {
  const r = await api<BookmarksList>("/api/me/bookmarks");
  return r.items;
};

File: packages/core/src/setvault_core/schemas/bookmarks.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BookmarkOut(BaseModel):
    id: str
    live_set_id: str
    live_set_slug: str | None = None
    live_set_title: str | None = None
    position_seconds: int | None = None
    label: str | None = None
    created_at: datetime


class BookmarkCreateIn(BaseModel):
    position_seconds: int | None = Field(default=None, ge=0)
    label: str | None = Field(default=None, max_length=255)


class BookmarksListOut(BaseModel):
    items: list[BookmarkOut]
